src/text-input/index.tsx

```diff
@@ -106,46 +100,81 @@ function formatAutocomplete(autocomplete: string | boolean | undefined): string
 	return autocomplete;
 }
 
-function patternDiffer(
-	previousProperty: string | undefined,
-	newProperty: string | RegExp | undefined
-): PropertyChangeRecord {
-	const value = newProperty instanceof RegExp ? newProperty.source : newProperty;
-	return {
-		changed: previousProperty !== value,
-		value
-	};
-}
+const factory = create({
+	theme,
+	icache,
+	validity,
+	focus,
+	diffProperty,
+	invalidator
+}).properties<TextInputProperties>();
 
-@theme(css)
-@diffProperty('pattern', patternDiffer)
-@diffProperty('leading', reference)
-@diffProperty('trailing', reference)
-export class TextInput extends ThemedMixin(FocusMixin(WidgetBase))<TextInputProperties> {
-	private _onInput(event: Event) {
-		event.stopPropagation();
-		this.properties.onValue &&
-			this.properties.onValue((event.target as HTMLInputElement).value);
-	}
+export const TextInput = factory(function TextInput({
+	middleware: { icache, theme, validity, focus, diffProperty, invalidator },
+	properties
+}) {
+	diffProperty('pattern', (previous: TextInputProperties, next: TextInputProperties) => {
+		const value = next.pattern instanceof RegExp ? next.pattern.source : next.pattern;
+		if (value !== previous.pattern) {
+			invalidator();
+		}
+	});
+	diffProperty(
+		'leading',
+		(previous: TextInputProperties, next: TextInputProperties) =>
+			previous.leading !== next.leading && invalidator()
+	);
+	diffProperty(
+		'trailing',
+		(previous: TextInputProperties, next: TextInputProperties) =>
+			previous.leading !== next.leading && invalidator()
+	);
 
-	private _onKeyDown(event: KeyboardEvent) {
-		event.stopPropagation();
-		this.properties.onKeyDown &&
-			this.properties.onKeyDown(event.which, () => {
-				event.preventDefault();
-			});
-	}
+	const themeCss = theme.classes(css);
+	const _uuid = icache.getOrSet('uuid', uuid());
```

Comment:
Would this be better as `icache.getOrSet('uuid', () => uuid());`? I don't know how expensive it is to create a UUID, but this is going to run every render as-is.